exporters/common/src/main/java/io/opentelemetry/exporter/internal/ExporterStatusException.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.exporter.internal;
+
+import javax.annotation.Nullable;
+
+/**
+ * An exception that records a status code from a HTTP or GRPC exporter, along with a root cause
+ * throwable.
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+public final class ExporterStatusException extends Exception {
```

Comment:
I think we need to be able to have different representations for http and grpc errors, and for exports that fail with a response from the server vs. exports that fail exceptionally. I sketched out a exception class hierarchy. What do you think?

```
package io.opentelemetry.exporter.internal;

import io.opentelemetry.exporter.internal.grpc.GrpcResponse;
import io.opentelemetry.exporter.internal.http.HttpSender;
import javax.annotation.Nullable;

public abstract class FailedExportException extends Exception {

  private FailedExportException(@Nullable Throwable cause) {
    super(cause);
  }

  /** Indicates an HTTP export failed after receiving a response from the server. */
  public static HttpExportException httpFailedWithResponse(HttpSender.Response response) {
    return new HttpExportException(response, null);
  }

  /** Indicates an HTTP export failed exceptionally without receiving a response from the server. */
  public static HttpExportException httpFailedExceptionally(Throwable cause) {
    return new HttpExportException(null, cause);
  }

  /** Indicates an HTTP export failed after receiving a response from the server. */
  public static GrpcExportException grpcFailedWithResponse(GrpcResponse response) {
    return new GrpcExportException(response, null);
  }

  /** Indicates an HTTP export failed exceptionally without receiving a response from the server. */
  public static GrpcExportException grpcFailedExceptionally(Throwable cause) {
    return new GrpcExportException(null, cause);
  }

  /** Returns true if the export failed with a response from the server. */
  abstract boolean failedWithResponse();

  /** Represents the failure of an HTTP exporter. */
  public static final class HttpExportException extends FailedExportException {

    @Nullable private final HttpSender.Response response;
    @Nullable private final Throwable cause;

    private HttpExportException(@Nullable HttpSender.Response response, @Nullable Throwable cause) {
      super(cause);
      this.response = response;
      this.cause = cause;
    }

    @Override
    boolean failedWithResponse() {
      return response != null;
    }

    /**
     * Returns the response if the export failed with a response from the server, or null if the
     * export failed exceptionally with no response.
     */
    @Nullable
    public HttpSender.Response getResponse() {
      return response;
    }

    /**
     * Returns the exceptional cause of failure, or null if the export failed with a response from
     * the server.
     */
    @Nullable
    public Throwable getCause() {
      return cause;
    }
  }

  /** Represents the failure of a gRPC exporter. */
  public static final class GrpcExportException extends FailedExportException {

    @Nullable private final GrpcResponse response;
    @Nullable private final Throwable cause;

    private GrpcExportException(@Nullable GrpcResponse response, @Nullable Throwable cause) {
      super(cause);
      this.response = response;
      this.cause = cause;
    }

    @Override
    boolean failedWithResponse() {
      return response != null;
    }

    /**
     * Returns the response if the export failed with a response from the server, or null if the
     * export failed exceptionally with no response.
     */
    @Nullable
    public GrpcResponse getResponse() {
      return response;
    }

    /**
     * Returns the exceptional cause of failure, or null if the export failed with a response from
     * the server.
     */
    @Nullable
    public Throwable getCause() {
      return cause;
    }
  }
}
```

Reply:
This is great actually, and gives a lot more flexibility to clients.